tests/c/CAsyncTest.php

```diff
@@ -17,7 +17,7 @@
 final class CAsyncTest extends HackTest {
 
   public static function provideTestGenFirst(
-  ): vec<(Awaitable<Traversable<mixed>>, mixed)> {
+  ): vec<(Awaitable<Traversable<mixed>>, ?nonnull)> {
```

Comment:
The typehint on the method that relies on this dataprovider is nullable.
The linter is not acutally checking assignability and matches things on a token level.

In `mixed` assign to `?T`, `mixed` tries to match `?`, which is a failure.
In `?nonnull` assign to `?T`, `?` matches `?` and `mixed` assigns to `T`, since generics are treated as wildcards.

I think to more preferable change would be for me to change the test function itself to take `mixed` and revert `?nonnull` typehints to `mixed`.

Reply:
Could I make this change without breaking your import into phabricator?

Reply:
I think the right fix here would be to add a special case to the linter so that it understands that mixed can match a nullable type (mixed is the only type that can, so this would be just that one special case).

The function is typed as generic `?T`, I don't think we want to change that to `mixed`.

> Could I make this change without breaking your import into phabricator?

Don't worry about making changes here, it's easier for me to change simple things like this than to re-import.

Reply:
> The function is typed as generic ?T, I don't think we want to change that to mixed.

Heh, I just realized this comment was totally wrong, in this particular case we probably should change it to `mixed` (but probably better to leave that for a separate PR):

```hack
  public function testFirstAsync<T>(
    Awaitable<Traversable<T>> $awaitable,
    ?T $expected,
  ): void {
```

The `T` here is useless, if you pass in `$awaitable` and `$expected` with completely different types, Hack will just resolve `T = mixed` and happily accept it. So pretending that the generics here give us any type safety is just misleading.

That said, the rest of the comment still applies -- we should still fix `?nonnull` vs `mixed` since there are other similar functions where generics are OK.

---

In case you're interested, the exact criterion for when generics are useless is:

- if a generic param is used only in contravariant positions (e.g. function arguments) it is equivalent to `mixed`
- if a generic param is used only in covariant positions (e.g. function return type, or arguments of a function that is itself an argument) it is equivalent to `nothing`

You can add `check_redundant_generics=true` to your .hhconfig to detect the latter (not that common, but it found a few cases in HSL when I tried), I think it's supposed to eventually also detect the former (which is more common).

I vaguely remember hearing that there might be some plans to fix this in the future (make it possible to declare a function like `testFirstAsync` above but where the generics actually enforce something), so removing them from HSL might be premature.

Reply:
> In case you're interested, the exact criterion for when generics are useless is

I am very much interested in this.
I heard Facebook has a linter that enforces that you can't have provably useless generics.